### PR TITLE
Add `torch.swapdims` and `torch.swapaxes`

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference #38349 

Delegates to `torch.transpose` (not sure what is the best way to alias)

TODO:
* [x] Add test
* [x] Add documentation

### CODE REVIEW

File: aten/src/ATen/core/aten_interned_strings.h [hunk 1]
```diff
@@ -657,6 +657,7 @@ _(aten, rsub) \
 _(aten, sum) \
 _(aten, sum_to_size) \
 _(aten, svd) \
+_(aten, swapdim) \
```
Comment (mruberry) [inline aten/src/ATen/core/aten_interned_strings.h:660]: names should be "swapdims" and "swapaxes" (since "axes" is plural)

File: aten/src/ATen/native/TensorShape.cpp [hunk 2]
```diff
@@ -1939,4 +1939,8 @@ Tensor movedim(const Tensor& self, int64_t src, int64_t dst) {
   return at::movedim(self, IntArrayRef{src}, IntArrayRef{dst});
 }
 
+Tensor swapdim(const Tensor& self, int64_t dim1, int64_t dim2) {
```
Comment (mruberry) [inline aten/src/ATen/native/TensorShape.cpp:1942]: transpose also has an inplace version: https://github.com/pytorch/pytorch/blob/a814231616f5db7fcdb8793f64c12d652d1e8572/aten/src/ATen/native/native_functions.yaml#L3512

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -5699,6 +5699,10 @@
     CPU: _svd_helper_cpu
     CUDA: _svd_helper_cuda
 
+- func: swapdim(Tensor(a) self, int dim1, int dim2) -> Tensor(a)
```
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:5702]: "swapdims"

Please also add "swapaxes," both as aliases of transpose. Instructions for adding aliases are here:

https://github.com/pytorch/pytorch/blob/a814231616f5db7fcdb8793f64c12d652d1e8572/aten/src/ATen/native/native_functions.yaml#L232

Comment (mruberry) [reply aten/src/ATen/native/native_functions.yaml:5702]: (You don't have to implement the aliases for the named variant of transpose).

File: docs/source/tensor_view.rst [hunk 4]
```diff
@@ -73,6 +73,7 @@ For reference, here’s a full list of view ops in PyTorch:
 - :meth:`~torch.Tensor.unbind`
 - :meth:`~torch.Tensor.split`
 - :meth:`~torch.Tensor.split_with_sizes`
+- :meth:`~torch.Tensor.swapdim`
```
Comment (mruberry) [inline docs/source/tensor_view.rst:76]: Also add entries to tensor.rst, since swapdims and swapaxes will be methods, too

File: test/test_torch.py [hunk 5]
```diff
@@ -19024,6 +19024,46 @@ def make_index_negative(sequence, idx):
         np_fn = partial(np.moveaxis, source=(), destination=())
         self.compare_with_numpy(torch_fn, np_fn, x, device=None, dtype=None)
 
+    @dtypes(torch.int64, torch.float, torch.complex128)
+    def test_swapdim_invalid(self, device, dtype):
```
Comment (mruberry) [inline test/test_torch.py:19028]: For both of these tests use "transpose" in the names and start by iterating over the functions:

```
for fn in (torch.transpose, torch.swapdims, torch.swapaxes):
  ...
```

File: test/test_torch.py [hunk 6]
```diff
@@ -19024,6 +19024,46 @@ def make_index_negative(sequence, idx):
...
+        x = self._generate_input(shape, dtype, device, False)
+
+        # Invalid `source` and `destination` dimension
+        with self.assertRaisesRegex(IndexError, "Dimension out of range"):
+            torch.swapdim(x, 5, 0)
+
+        with self.assertRaisesRegex(IndexError, "Dimension out of range"):
+            torch.swapdim(x, 0, 5)
+
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
```
Comment (mruberry) [inline test/test_torch.py:19039]: We now require NumPy to run the test suite and these skips are no longer needed.

File: test/test_torch.py [hunk 7]
```diff
@@ -19024,6 +19024,46 @@ def make_index_negative(sequence, idx):
...
+    @dtypes(torch.int64, torch.float, torch.complex128)
+    def test_swapdim(self, device, dtype):
```
Comment (mruberry) [inline test/test_torch.py:19041]: "test_transpose_vs_numpy", see note above

File: test/test_torch.py [hunk 8]
```diff
@@ -19996,9 +20036,9 @@ def test_split_view(self, device):
             v[0, 0] = idx + 1
             self.assertEqual(t[idx, 0], v[0, 0])
 
-    def test_movedim_view(self, device):
+    def _test_movedim_swapdim_view(self, device, op):
```
Comment (mruberry) [inline test/test_torch.py:20039]: We have test_tranpose_view, already, which you can reuse with the above technique.

File: tools/autograd/derivatives.yaml [hunk 9]
```diff
@@ -1020,6 +1020,9 @@
 - name: symeig(Tensor self, bool eigenvectors=False, bool upper=True) -> (Tensor eigenvalues, Tensor eigenvectors)
   self: symeig_backward(grads, self, eigenvectors, upper, eigenvalues, eigenvectors_return)
 
+- name: swapdim(Tensor(a) self, int dim1, int dim2) -> Tensor(a)
```
Comment (mruberry) [inline tools/autograd/derivatives.yaml:1023]: You won't need to add a derivatives.yaml entry if this is implemented using the alias instructions.

File: tools/autograd/gen_autograd.py [hunk 10]
```diff
@@ -87,7 +87,7 @@
 RETURNS_VIEWS_OF_INPUT = set(VIEW_FUNCTIONS.keys()).union({
     'chunk', 'detach', 'contiguous', 'reshape', 'reshape_as',
     'expand_as', 'view_as', 'real', 'imag', 'narrow', 'movedim',
-    'tensor_split'
+    'tensor_split', 'swapdim'
```
Comment (mruberry) [inline tools/autograd/gen_autograd.py:90]: +swapaxes

File: torch/_torch_docs.py [hunk 11]
```diff
@@ -5331,6 +5331,40 @@ def merge_dicts(*dicts):
             [[-0.8437,  0.1727, -0.1398]]])
 """.format(**common_args))
 
+add_docstr(torch.swapdim, r"""
+swapdim(input, dim1, dim2) -> Tensor
+
+Swaps the dimension of :attr:`input` at the position in :attr:`dim1`
```
Comment (mruberry) [inline torch/_torch_docs.py:5337]: This should just say it's an alias for transpose and link to that documentation.

**Comment by mruberry** [comment]
Hey @kshitij12345, thanks for this PR! I added a link to the instructions for adding an alias. Would you go through them and also add the "swapaxes" alias?

File: aten/src/ATen/core/aten_interned_strings.h (hunk 1 above)
Comment (kshitij12345) [reply aten/src/ATen/core/aten_interned_strings.h:660]: Ah right. Thanks!

**Comment by kshitij12345** [comment]
@mruberry Thanks for the review and reference to adding an alias.

I just had 2 questions.
1. Just wanted to confirm that the functions would have following signatures 
```python
swapdims(Tensor(a) self, int dim1, int dim2) -> Tensor(a)
swapaxes(Tensor(a) self, int axis1, int axis2) -> Tensor(a) # Note the parameter names are different.
```

2. Should we really link swapaxes/swapdims as an alias to transpose?
The `torch.transpose` API is different from `numpy.transpose` and thus it is direct alias for our case.
Reference: 
https://numpy.org/doc/stable/reference/generated/numpy.transpose.html
https://pytorch.org/docs/stable/generated/torch.transpose.html
So I feel that `swapdims` can have it's own documentation instead of pointing to `transpose`'s doc and `swapaxes` can be linked to `swapdims`'s doc

Do let me know how you feel about the same.

**Comment by kshitij12345** [comment]
Also I am getting the following failures when I run the alias test locally.
Note that it does pass for inplace variants.

Thanks!
```
============================================================================== FAILURES ==============================================================================
________________________________________________ TestOpNormalizationCPU.test_jit_op_alias_normalization_swapaxes_cpu _________________________________________________

self = <test_op_aliases.TestOpNormalizationCPU testMethod=test_jit_op_alias_normalization_swapaxes_cpu>, device = 'cpu'
info = <test_op_aliases.AliasInfo object at 0x7f9ef17dc048>

    def _test_jit_op_alias_normalization(self, device, info=info):
        tensor = torch.tensor
        op = info.alias_op
        is_inplace = info.alias_name.endswith('_')
    
        # Checks that scripting converts aliases
        # NOTE: the code to test scripting must be generated since
        #   scripting does not support splatting args or directly
        #   calling torch.Tensor methods. The following
        #   splats args after the first tensor by inlining them as constants.
        if is_inplace:
            fn_template = '''
                def _fn(t):
                    return t.{alias_name}({args})
            '''
            arg_string = ', '.join((str(arg) for arg in info.get_args(device)))
            script = fn_template.format(alias_name=info.alias_name, args=arg_string)
        else:
            fn_template = '''
                def _fn(t):
                    return op(t{args})
            '''
            arg_string = ", " + ', '.join((str(arg) for arg in info.get_args(device)))
            script = fn_template.format(args=arg_string)
    
        # Compiles script
>       scripted = torch.jit.CompilationUnit(script)._fn

test/test_op_aliases.py:203: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../PytorchENV/lib/python3.7/site-packages/torch/jit/_script.py:1063: in __init__
    self.define(lang, _frames_up=_frames_up + 1)
../PytorchENV/lib/python3.7/site-packages/torch/jit/_script.py:1068: in define
    self._c.define(lang, rcb)
../PytorchENV/lib/python3.7/site-packages/torch/_jit_internal.py:845: in _try_get_dispatched_fn
    return boolean_dispatched.get(fn)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <WeakKeyDictionary at 0x7f9ef9af9048>, key = <method 'swapaxes' of 'torch._C._TensorBase' objects>, default = None

    def get(self, key, default=None):
>       return self.data.get(ref(key),default)
E       TypeError: cannot create weak reference to 'method_descriptor' object

../PytorchENV/lib/python3.7/weakref.py:433: TypeError
________________________________________________ TestOpNormalizationCPU.test_jit_op_alias_normalization_swapdims_cpu _________________________________________________

self = <test_op_aliases.TestOpNormalizationCPU testMethod=test_jit_op_alias_normalization_swapdims_cpu>, device = 'cpu'
info = <test_op_aliases.AliasInfo object at 0x7f9ef17d9f48>

    def _test_jit_op_alias_normalization(self, device, info=info):
        tensor = torch.tensor
        op = info.alias_op
        is_inplace = info.alias_name.endswith('_')
    
        # Checks that scripting converts aliases
        # NOTE: the code to test scripting must be generated since
        #   scripting does not support splatting args or directly
        #   calling torch.Tensor methods. The following
        #   splats args after the first tensor by inlining them as constants.
        if is_inplace:
            fn_template = '''
                def _fn(t):
                    return t.{alias_name}({args})
            '''
            arg_string = ', '.join((str(arg) for arg in info.get_args(device)))
            script = fn_template.format(alias_name=info.alias_name, args=arg_string)
        else:
            fn_template = '''
                def _fn(t):
                    return op(t{args})
            '''
            arg_string = ", " + ', '.join((str(arg) for arg in info.get_args(device)))
            script = fn_template.format(args=arg_string)
    
        # Compiles script
>       scripted = torch.jit.CompilationUnit(script)._fn

test/test_op_aliases.py:203: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../PytorchENV/lib/python3.7/site-packages/torch/jit/_script.py:1063: in __init__
    self.define(lang, _frames_up=_frames_up + 1)
../PytorchENV/lib/python3.7/site-packages/torch/jit/_script.py:1068: in define
    self._c.define(lang, rcb)
../PytorchENV/lib/python3.7/site-packages/torch/_jit_internal.py:845: in _try_get_dispatched_fn
    return boolean_dispatched.get(fn)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <WeakKeyDictionary at 0x7f9ef9af9048>, key = <method 'swapdims' of 'torch._C._TensorBase' objects>, default = None

    def get(self, key, default=None):
>       return self.data.get(ref(key),default)
E       TypeError: cannot create weak reference to 'method_descriptor' object

../PytorchENV/lib/python3.7/weakref.py:433: TypeError
====================================================================== short test summary info =======================================================================
FAILED test/test_op_aliases.py::TestOpNormalizationCPU::test_jit_op_alias_normalization_swapaxes_cpu - TypeError: cannot create weak reference to 'method_descripto...
FAILED test/test_op_aliases.py::TestOpNormalizationCPU::test_jit_op_alias_normalization_swapdims_cpu - TypeError: cannot create weak reference to 'method_descripto...
============================================================= 2 failed, 126 passed, 48 skipped in 5.34s ==============================================================
```

**Comment by mruberry** [comment]
> @mruberry Thanks for the review and reference to adding an alias.
> 
> I just had 2 questions.
> 
> 1. Just wanted to confirm that the functions would have following signatures
> 
> ```python
> swapdims(Tensor(a) self, int dim1, int dim2) -> Tensor(a)
> swapaxes(Tensor(a) self, int axis1, int axis2) -> Tensor(a) # Note the parameter names are different.
> ```

This is a neat idea. Sure, let's try it.

> 2. Should we really link swapaxes/swapdims as an alias to transpose?
>    The `torch.transpose` API is different from `numpy.transpose` and thus it is direct alias for our case.
>    Reference:
>    https://numpy.org/doc/stable/reference/generated/numpy.transpose.html
>    https://pytorch.org/docs/stable/generated/torch.transpose.html
>    So I feel that `swapdims` can have it's own documentation instead of pointing to `transpose`'s doc and `swapaxes` can be linked to `swapdims`'s doc

What would you think about aliasing them but adding a comment about how swapdims/swapaxes are analogous to NumPy's swapaxes function? Would that be the best of both worlds?

File: test/test_op_aliases.py [hunk 12]
```diff
@@ -150,6 +150,18 @@ def __init__(self,
     AliasInfo('true_divide_', torch.Tensor.true_divide_, 'div_', torch.Tensor.div_,
               lambda d: torch.randn(20, device=d), get_args=lambda d: (torch.rand(20, device=d) + .1,),
               decorators=(onlyCPU,)),
+    AliasInfo('swapdims', torch.Tensor.swapdims, 'transpose', torch.Tensor.transpose,
+              lambda d: torch.randn(20, 3, 2, 1, device=d), get_args=lambda d: (3, 1),
+              decorators=(onlyCPU,)),
```
Comment (mruberry) [inline test/test_op_aliases.py:155]: What happens when this decorator is removed? Currently we run some of these tests only on CPU to avoid a memory leak issue with CUDA tensors, but I think that only applies if an argument (from get_args) is a tensor, which shouldn't be the case here.

**Comment by mruberry** [comment]
> Also I am getting the following failures when I run the alias test locally.
> Note that it does pass for inplace variants.
> 
> Thanks!
> 
> ```
> ============================================================================== FAILURES ==============================================================================
> ________________________________________________ TestOpNormalizationCPU.test_jit_op_alias_normalization_swapaxes_cpu _________________________________________________
> 
> self = <test_op_aliases.TestOpNormalizationCPU testMethod=test_jit_op_alias_normalization_swapaxes_cpu>, device = 'cpu'
> info = <test_op_aliases.AliasInfo object at 0x7f9ef17dc048>
> 
>     def _test_jit_op_alias_normalization(self, device, info=info):
>         tensor = torch.tensor
>         op = info.alias_op
>         is_inplace = info.alias_name.endswith('_')
>     
>         # Checks that scripting converts aliases
>         # NOTE: the code to test scripting must be generated since
>         #   scripting does not support splatting args or directly
>         #   calling torch.Tensor methods. The following
>         #   splats args after the first tensor by inlining them as constants.
>         if is_inplace:
>             fn_template = '''
>                 def _fn(t):
>                     return t.{alias_name}({args})
>             '''
>             arg_string = ', '.join((str(arg) for arg in info.get_args(device)))
>             script = fn_template.format(alias_name=info.alias_name, args=arg_string)
>         else:
>             fn_template = '''
>                 def _fn(t):
>                     return op(t{args})
>             '''
>             arg_string = ", " + ', '.join((str(arg) for arg in info.get_args(device)))
>             script = fn_template.format(args=arg_string)
>     
>         # Compiles script
> >       scripted = torch.jit.CompilationUnit(script)._fn
> 
> test/test_op_aliases.py:203: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> ../PytorchENV/lib/python3.7/site-packages/torch/jit/_script.py:1063: in __init__
>     self.define(lang, _frames_up=_frames_up + 1)
> ../PytorchENV/lib/python3.7/site-packages/torch/jit/_script.py:1068: in define
>     self._c.define(lang, rcb)
> ../PytorchENV/lib/python3.7/site-packages/torch/_jit_internal.py:845: in _try_get_dispatched_fn
>     return boolean_dispatched.get(fn)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> self = <WeakKeyDictionary at 0x7f9ef9af9048>, key = <method 'swapaxes' of 'torch._C._TensorBase' objects>, default = None
> 
>     def get(self, key, default=None):
> >       return self.data.get(ref(key),default)
> E       TypeError: cannot create weak reference to 'method_descriptor' object
> 
> ../PytorchENV/lib/python3.7/weakref.py:433: TypeError
> ________________________________________________ TestOpNormalizationCPU.test_jit_op_alias_normalization_swapdims_cpu _________________________________________________
> 
> self = <test_op_aliases.TestOpNormalizationCPU testMethod=test_jit_op_alias_normalization_swapdims_cpu>, device = 'cpu'
> info = <test_op_aliases.AliasInfo object at 0x7f9ef17d9f48>
> 
>     def _test_jit_op_alias_normalization(self, device, info=info):
>         tensor = torch.tensor
>         op = info.alias_op
>         is_inplace = info.alias_name.endswith('_')
>     
>         # Checks that scripting converts aliases
>         # NOTE: the code to test scripting must be generated since
>         #   scripting does not support splatting args or directly
>         #   calling torch.Tensor methods. The following
>         #   splats args after the first tensor by inlining them as constants.
>         if is_inplace:
>             fn_template = '''
>                 def _fn(t):
>                     return t.{alias_name}({args})
>             '''
>             arg_string = ', '.join((str(arg) for arg in info.get_args(device)))
>             script = fn_template.format(alias_name=info.alias_name, args=arg_string)
>         else:
>             fn_template = '''
>                 def _fn(t):
>                     return op(t{args})
>             '''
>             arg_string = ", " + ', '.join((str(arg) for arg in info.get_args(device)))
>             script = fn_template.format(args=arg_string)
>     
>         # Compiles script
> >       scripted = torch.jit.CompilationUnit(script)._fn
> 
> test/test_op_aliases.py:203: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> ../PytorchENV/lib/python3.7/site-packages/torch/jit/_script.py:1063: in __init__
>     self.define(lang, _frames_up=_frames_up + 1)
> ../PytorchENV/lib/python3.7/site-packages/torch/jit/_script.py:1068: in define
>     self._c.define(lang, rcb)
> ../PytorchENV/lib/python3.7/site-packages/torch/_jit_internal.py:845: in _try_get_dispatched_fn
>     return boolean_dispatched.get(fn)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> self = <WeakKeyDictionary at 0x7f9ef9af9048>, key = <method 'swapdims' of 'torch._C._TensorBase' objects>, default = None
> 
>     def get(self, key, default=None):
> >       return self.data.get(ref(key),default)
> E       TypeError: cannot create weak reference to 'method_descriptor' object
> 
> ../PytorchENV/lib/python3.7/weakref.py:433: TypeError
> ====================================================================== short test summary info =======================================================================
> FAILED test/test_op_aliases.py::TestOpNormalizationCPU::test_jit_op_alias_normalization_swapaxes_cpu - TypeError: cannot create weak reference to 'method_descripto...
> FAILED test/test_op_aliases.py::TestOpNormalizationCPU::test_jit_op_alias_normalization_swapdims_cpu - TypeError: cannot create weak reference to 'method_descripto...
> ============================================================= 2 failed, 126 passed, 48 skipped in 5.34s ==============================================================
> ```

Unfortunately I'm not sure what's going on here. I think you'll have to start debugging, sorry.

**Comment by kshitij12345** [comment]
> What would you think about aliasing them but adding a comment about how swapdims/swapaxes are analogous to NumPy's swapaxes function? Would that be the best of both worlds?

Sure. That makes sense.

**Comment by kshitij12345** [comment]
https://github.com/pytorch/pytorch/issues/47358#issuecomment-722252021
> Issue looks to arise from use of `abs`,
> 
> ```python
> import numpy as np
> import torch
> 
> print(torch.__version__)
> x = torch.tensor([0.-1.0000e+20j] * 10)
> print("ABS:", x.abs())
> print("ABS:", np.abs(x.numpy()))
> ```
> 
> Output
> 
> ```
> 1.8.0a0+f19637e
> ABS: tensor([       inf,        inf,        inf,        inf,        inf,        inf,
>                inf,        inf, 1.0000e+20, 1.0000e+20])
> ABS: [1.e+20 1.e+20 1.e+20 1.e+20 1.e+20 1.e+20 1.e+20 1.e+20 1.e+20 1.e+20]
> ```



> Issue looks to arise from use of `abs`,
> 
> ```python
> import numpy as np
> import torch
> 
> print(torch.__version__)
> x = torch.tensor([0.-1.0000e+20j] * 10)
> print("ABS:", x.abs())
> print("ABS:", np.abs(x.numpy()))
> ```
> 
> Output
> 
> ```
> 1.8.0a0+f19637e
> ABS: tensor([       inf,        inf,        inf,        inf,        inf,        inf,
>                inf,        inf, 1.0000e+20, 1.0000e+20])
> ABS: [1.e+20 1.e+20 1.e+20 1.e+20 1.e+20 1.e+20 1.e+20 1.e+20 1.e+20 1.e+20]
> ```

**Comment by kshitij12345** [comment]
@JackCaoG 

XLA failure looks real.

Should I open an issue on xla repo for the same?
```
======================================================================
Nov 05 14:55:32 ERROR [0.027s]: test_transpose_invalid_xla_float32 (__main__.TestTorchDeviceTypeXLA)
Nov 05 14:55:32 ----------------------------------------------------------------------
Nov 05 14:55:32 Traceback (most recent call last):
Nov 05 14:55:32   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 272, in instantiated_test
Nov 05 14:55:32     result = test_fn(self, *args)
Nov 05 14:55:32   File "/var/lib/jenkins/workspace/xla/test/../../test/test_torch.py", line 19441, in test_transpose_invalid
Nov 05 14:55:32     fn(x, 5, 0)
Nov 05 14:55:32 RuntimeError: torch_xla/csrc/helpers.cpp:97 : Check failed: min_shape_dim <= dim && dim <= max_shape_dim 
Nov 05 14:55:32 *** Begin stack trace ***
Nov 05 14:55:32 	tensorflow::CurrentStackTrace[abi:cxx11]()
Nov 05 14:55:32 	torch_xla::XlaHelpers::GetCanonicalDimensionIndex(long long, long long)
Nov 05 14:55:32 	torch_xla::XlaHelpers::MakeTransposePermutation(long long, long long, long long)
Nov 05 14:55:32 	torch_xla::XLATensor::transpose(torch_xla::XLATensor const&, long long, long long)
Nov 05 14:55:32 	torch_xla::AtenXlaType::transpose(at::Tensor const&, long, long)
Nov 05 14:55:32 	c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoRuntimeFunctor_<at::Tensor (*)(at::Tensor const&, long, long), at::Tensor, c10::guts::typelist::typelist<at::Tensor const&, long, long> >, at::Tensor (at::Tensor const&, long, long)>::call(c10::OperatorKernel*, at::Tensor const&, long, long)
Nov 05 14:55:32
```

**Comment by JackCaoG** [comment]
@kshitij12345 Yup, please open an github issue on pt/xla, we will try to take a look soon!

**Comment by kshitij12345** [comment]
@JackCaoG Thanks for the quick reply! Will open one.

**Comment by mruberry** [comment]
The test on XLA can also be temporarily disabled using the `@onlyOnCPUAndCUDA` decorator.

**Comment by kshitij12345** [comment]
@mruberry PTAL :)

File: test/test_torch.py [hunk 13]
```diff
@@ -20411,9 +20465,9 @@ def test_split_view(self, device):
             v[0, 0] = idx + 1
             self.assertEqual(t[idx, 0], v[0, 0])
 
-    def test_movedim_view(self, device):
+    def _test_movedim_view(self, device, op):
```
Comment (walterddr) [inline test/test_torch.py:20468]: is this change relevant to `swapdims` & `swapaxes`?

Comment (kshitij12345) [reply test/test_torch.py:20468]: Not really relevant to `swapdims/swapaxes`. Was using it previously, but right now it only tests `movedim`. (Test is moved to test_transpose_view)

Have kept this particular change since it adds one more test case for `movedim`

File: aten/src/ATen/native/native_functions.yaml [hunk 14]
```diff
@@ -6211,6 +6211,26 @@
...
 
+# swapaxes, alias for transpose
+- func: swapaxes(Tensor(a) self, int axis1, int axis2) -> Tensor(a)
+  use_c10_dispatcher: full
+  variants: function, method
+
+- func: swapaxes_(Tensor(a!) self, int axis1, int axis2) -> Tensor(a!)
+  use_c10_dispatcher: full
+  variants: method
+  device_guard: False
```
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:6222]: If we look at the dispatch entries for transpose:

```
- func: transpose.int(Tensor(a) self, int dim0, int dim1) -> Tensor(a)
  use_c10_dispatcher: full
  variants: function, method
  device_guard: False
  dispatch:
    DefaultBackend: transpose

- func: transpose_(Tensor(a!) self, int dim0, int dim1) -> Tensor(a!)
  use_c10_dispatcher: full
  variants: method
  device_guard: False
  dispatch:
    DefaultBackend: transpose_
```

We see that `device_guard: False` appears on both the in place and out-of-place variants and that the dispatch is `DefaultBackend: transpose` or `DefaultBackend: transpose_`. 

Note also that transpose starts its dim indexing at zero, so swapaxes should have axis0 and axis1, and swapdims should have dim0 and dim1, too.

File: test/test_torch.py (hunk 13 above)
Comment (mruberry) [reply test/test_torch.py:20461]: Rename this to something like `run_test` or `helper` and define it within `test_movedim_view`.

File: torch/_tensor_docs.py [hunk 15]
```diff
@@ -3536,6 +3536,31 @@ def callable(a, b) -> number
 See :func:`torch.symeig`
 """)
 
+add_docstr_all('swapdims', r"""
+swapdims(dim1, dim2) -> Tensor
```
Comment (mruberry) [inline torch/_tensor_docs.py:3540]: Numbering on dims and axes will need an update here.

File: torch/_torch_docs.py (hunk 11 above)
Comment (mruberry) [reply torch/_torch_docs.py:5635]: This section of the documentation still needs to be deleted. See, for example, the documentation for subtract: https://pytorch.org/docs/master/generated/torch.subtract.html#torch-subtract.

File: torch/_torch_docs.py [hunk 16]
```diff
@@ -5627,6 +5627,78 @@ def merge_dicts(*dicts):
             [[-0.8437,  0.1727, -0.1398]]])
 """.format(**common_args))
 
+add_docstr(torch.swapdims, r"""
+swapdims(input, dim1, dim2) -> Tensor
```
Comment (mruberry) [inline torch/_torch_docs.py:5631]: The numbering on the dim arguments will need an update

File: torch/_torch_docs.py [hunk 17]
```diff
@@ -5627,6 +5627,78 @@ def merge_dicts(*dicts):
...
+    >>> torch.swapdims(x, 0, 2)
+    tensor([[[0, 4],
+            [2, 6]],
+
+            [[1, 5],
+            [3, 7]]])    
+""".format(**common_args))
+
+add_docstr(torch.swapaxes, r"""
+swapaxes(input, axis1, axis2) -> Tensor
```
Comment (mruberry) [inline torch/_torch_docs.py:5667]: The numbering on the axis arguments will need an update

File: torch/_torch_docs.py [hunk 18]
```diff
@@ -5627,6 +5627,78 @@ def merge_dicts(*dicts):
...
+
+Alias for :func:`torch.transpose`.
+
+Swaps the dimension of :attr:`input` at the position in :attr:`axis1`
```
Comment (mruberry) [inline torch/_torch_docs.py:5671]: This part of the documentation still needs to be deleted.

File: torch/overrides.py [hunk 19]
```diff
@@ -739,6 +739,8 @@ def get_testing_overrides() -> Dict[Callable, Callable]:
         torch.svd: lambda input, some=True, compute_uv=True, out=None: -1,
         torch.svd_lowrank: lambda input, q=6, niter=2, M=None: -1,
         torch.symeig: lambda input, eigenvectors=False, upper=True, out=None: -1,
+        torch.swapaxes: lambda input, dim1, dim2: -1,
```
Comment (mruberry) [inline torch/overrides.py:742]: The numbering here and on the next line will need to be updated

**Comment by mruberry** [comment]
> @kshitij12345 Yup, please open an github issue on pt/xla, we will try to take a look soon!

We can skip the tests on XLA for now. XLA can restore them once it implements the functions (on its own time).

**Review by mruberry — COMMENTED** [review]
Hey @kshitij12345! This looks very good and close to being finished. I made some cleanup comments that should be easy to address.

File: aten/src/ATen/native/native_functions.yaml (hunk 14 above)
Comment (kshitij12345) [reply aten/src/ATen/native/native_functions.yaml:6222]: @mruberry So should we remove `device_guard` from the signature? Sorry didn't understand comment.


> Note also that transpose starts its dim indexing at zero, so swapaxes should have axis0 and axis1, and swapdims should have dim0 and dim1, too.

Sure. The reason for this was because numpy names argument `axis1, axis2`, but makes sense to be more similar to transpose.

Comment (kshitij12345) [reply aten/src/ATen/native/native_functions.yaml:6222]: @mruberry I think I understand now. `device_guard` should be on both signatues. Thanks!

File: torch/_torch_docs.py (hunk 18 above)
Comment (kshitij12345) [reply torch/_torch_docs.py:5671]: > What would you think about aliasing them but adding a comment about how swapdims/swapaxes are analogous to NumPy's swapaxes function? Would that be the best of both worlds?

https://github.com/pytorch/pytorch/pull/46041#issuecomment-713005625

Isn't the plan to keep alias as well as some documentation explaining this behaviour?

**Comment by kshitij12345** [comment]
@mruberry PTAL:)

File: aten/src/ATen/native/native_functions.yaml (hunk 14 above)
Comment (mruberry) [reply aten/src/ATen/native/native_functions.yaml:6223]: Yep, this looks correct.

File: torch/_torch_docs.py (hunk 18 above)
Comment (mruberry) [reply torch/_torch_docs.py]: What about deleting the redundant parts and then, after the sentence explaining its an alias, adding a new sentence like:

"This function is equivalent to Numpy's swapaxes function." 

?

**Review by mruberry — COMMENTED** [review]
Hey @kshitij12345! Everything looks great. Just have a final question about simplifying the docs.

File: torch/_torch_docs.py (hunk 18 above)
Comment (kshitij12345) [reply torch/_torch_docs.py]: Makes sense.

File: torch/_torch_docs.py [hunk 20]
```diff
@@ -5668,13 +5662,7 @@ def merge_dicts(*dicts):
 
 Alias for :func:`torch.transpose`.
 
-Swaps the dimension of :attr:`input` at the position in :attr:`axis0`
```
Comment (mruberry) [inline torch/_torch_docs.py]: "Numpy" -> "NumPy" here and in the swapdims doc.

**Review by mruberry — APPROVED** [review]
Looks great!

Just a last very small change to write "Numpy" as "NumPy."

**Comment by kshitij12345** [comment]
Updated. Thanks for the great review as always :)

**Comment by mruberry** [comment]
> @mruberry has imported this pull request. If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/intern/diff/D25022816/).

You're welcome ;) Thanks for the PR!
